xds/internal/xdsclient/authority.go

```diff
@@ -50,70 +262,121 @@ func (a *authority) unref() int {
+			a.handleWatchTimerExpiry(state, fmt.Errorf("watch for resource %q of type %s timed out", resourceName, rType.TypeEnum().String()))
+		})
+		resources[resourceName] = state
+		a.sendDiscoveryRequest(rType, resources)
 	}
-}
+	// Always add the new watcher to the set of watchers.
+	state.watchers[watcher] = true
 
-func (a *authority) watchRouteConfig(routeName string, cb func(xdsresource.RouteConfigUpdate, error)) (cancel func()) {
-	first, cancelF := a.pubsub.WatchRouteConfig(routeName, cb)
-	if first {
-		a.controller.AddWatch(xdsresource.RouteConfigResource, routeName)
+	// If we have a cached copy of the resource, notify the new watcher.
+	if state.cache != nil {
+		a.logger.Debugf("Resource type %q with resource name %q found in cache: %s", rType.TypeEnum(), resourceName, state.cache.ToJSON())
+		resource := state.cache
+		a.serializer.Schedule(func(context.Context) { watcher.OnUpdate(resource) })
 	}
+
 	return func() {
-		if cancelF() {
-			a.controller.RemoveWatch(xdsresource.RouteConfigResource, routeName)
+		a.resourcesMu.Lock()
+		defer a.resourcesMu.Unlock()
+
+		// We already have a reference to the resourceState for this particular
+		// resource. Avoid indexing into the two-level map to figure this out.
+
+		// Delete this particular watcher from the list of watchers, so that its
+		// callback will not be invoked in the future.
+		state.wState = watchStateCanceled
+		delete(state.watchers, watcher)
+		if len(state.watchers) > 0 {
+			return
 		}
+
+		// There are no more watchers for this resource, delete the state
+		// associated with it, and instruct the transport to send a request
+		// which does not include this resource name.
+		delete(resources, resourceName)
+		a.sendDiscoveryRequest(rType, resources)
 	}
 }
 
-func (a *authority) watchCluster(clusterName string, cb func(xdsresource.ClusterUpdate, error)) (cancel func()) {
-	first, cancelF := a.pubsub.WatchCluster(clusterName, cb)
-	if first {
-		a.controller.AddWatch(xdsresource.ClusterResource, clusterName)
+func (a *authority) handleWatchTimerExpiry(state *resourceState, err error) {
+	a.resourcesMu.Lock()
+	defer a.resourcesMu.Unlock()
+
+	if state.wState == watchStateCanceled {
+		return
 	}
-	return func() {
-		if cancelF() {
-			a.controller.RemoveWatch(xdsresource.ClusterResource, clusterName)
-		}
+
+	state.wState = watchStateTimeout
+	for watcher := range state.watchers {
+		watcher := watcher
+		a.serializer.Schedule(func(context.Context) { watcher.OnError(err) })
 	}
 }
 
-func (a *authority) watchEndpoints(clusterName string, cb func(xdsresource.EndpointsUpdate, error)) (cancel func()) {
-	first, cancelF := a.pubsub.WatchEndpoints(clusterName, cb)
-	if first {
-		a.controller.AddWatch(xdsresource.EndpointsResource, clusterName)
-	}
-	return func() {
-		if cancelF() {
-			a.controller.RemoveWatch(xdsresource.EndpointsResource, clusterName)
-		}
+func (a *authority) sendDiscoveryRequest(rType xdsresource.Type, resources map[string]*resourceState) {
```

Comment:
Is it important that resourcesMu is held here?  (I think so; document or rename to `sendDiscoveryRequestLocked`?)  Also maybe check for other places accessing `resources` that may not be holding the mutex or may not be making it explicit.